### ISSUE TITLE
vnc connection vmi error encountered reading from client stream

### ISSUE THREAD

**Issue by biliwewe** [kubevirt/kubevirt#2805]
**Is this a BUG REPORT or FEATURE REQUEST?**:

> /kind bug

**What happened**:
Kubevirt upgraded from v0.17.0 to v0.22.0. Using vnc to connect vmi exception. The first connection to vnc can succeed. Unable to connect after disconnecting.

**Environment**:
- KubeVirt version (use `virtctl version`): v0.22.0
- Kubernetes version (use `kubectl version`): v1.14.6
- OS (e.g. from /etc/os-release): centos7
- Kernel (e.g. `uname -a`): 5.2.11-1.el7.elrepo.x86_64

**The following is the log I captured.**

```log
virt-handler.log:
{"component":"virt-handler","kind":"","level":"info","msg":"Websocket connection upgraded","name":"win2k3","namespace":"default","pos":"console.go:170","timestamp":"2019-09-18T11:27:10.764254Z","uid":"d64513cb-da03-11e9-b872-0050563d4d0a"}
{"component":"virt-handler","kind":"","level":"info","msg":"Connecting to proc/57105/root/var/run/kubevirt-private/d64513cb-da03-11e9-b872-0050563d4d0a/virt-vnc","name":"win2k3","namespace":"default","pos":"console.go:171","timestamp":"2019-09-18T11:27:10.764500Z","uid":"d64513cb-da03-11e9-b872-0050563d4d0a"}
{"component":"virt-handler","kind":"","level":"info","msg":"Connected to proc/57105/root/var/run/kubevirt-private/d64513cb-da03-11e9-b872-0050563d4d0a/virt-vnc","name":"win2k3","namespace":"default","pos":"console.go:181","timestamp":"2019-09-18T11:27:10.764691Z","uid":"d64513cb-da03-11e9-b872-0050563d4d0a"}
{"component":"virt-handler","kind":"","level":"error","msg":"error encountered reading from client (virt-api) websocket","name":"win2k3","namespace":"default","pos":"console.go:192","reason":"websocket: close 1006 (abnormal closure): unexpected EOF","timestamp":"2019-09-18T11:27:10.905375Z","uid":"d64513cb-da03-11e9-b872-0050563d4d0a"}
{"component":"virt-handler","kind":"","level":"error","msg":"Error in proxing websocket and unix socket","name":"win2k3","namespace":"default","pos":"console.go:201","reason":"websocket: close 1006 (abnormal closure): unexpected EOF","timestamp":"2019-09-18T11:27:10.905642Z","uid":"d64513cb-da03-11e9-b872-0050563d4d0a"}
ERROR: logging before flag.Parse: I0918 11:27:10.905808 89590 server.go:2979] http: response.WriteHeader on hijacked connection
{"component":"virt-handler","kind":"","level":"error","msg":"error encountered reading from unix socket","name":"win2k3","namespace":"default","pos":"console.go:186","reason":"read unix @-\u003e/var/run/kubevirt-private/d64513cb-da03-11e9-b872-0050563d4d0a/virt-vnc: use of closed network connection","timestamp":"2019-09-18T11:27:10.906021Z","uid":"d64513cb-da03-11e9-b872-0050563d4d0a"}
```

```log
virt-api .log:
{"component":"virt-api","kind":"VirtualMachineInstance","level":"error","msg":"error encountered reading from client stream","name":"win2k3","namespace":"default","pos":"subresource.go:150","reason":"websocket: close 1001 (going away)","timestamp":"2019-09-18T11:23:23.519688Z","uid":"d64513cb-da03-11e9-b872-0050563d4d0a"}
{"component":"virt-api","kind":"VirtualMachineInstance","level":"error","msg":"Error in websocket proxy","name":"win2k3","namespace":"default","pos":"subresource.go:158","reason":"websocket: close 1001 (going away)","timestamp":"2019-09-18T11:23:23.519978Z","uid":"d64513cb-da03-11e9-b872-0050563d4d0a"}
ERROR: logging before flag.Parse: I0918 11:23:23.520170 1 server.go:2979] http: response.WriteHeader on hijacked connection
{"component":"virt-api","kind":"VirtualMachineInstance","level":"error","msg":"error encountered reading from virt-handler stream","name":"win2k3","namespace":"default","pos":"subresource.go:144","reason":"read tcp 10.42.0.10:41604-\u003e10.42.0.13:8186: use of closed network connection","timestamp":"2019-09-18T11:23:23.520557Z","uid":"d64513cb-da03-11e9-b872-0050563d4d0a"}

```
I am sure there is no problem in version v0.17.0, I can connect via vnc

**Comment by rmohr** [comment]
@ahadas can you have a look? It really looks like with the vnc/console refactoring something is broken here.

**Comment by biliwewe** [comment]
hi @ahadas can you help me?

**Comment by rmohr** [comment]
@biliwewe we have had multiple reports about this issue. It is for sure real. I had so far issues reproducing it myself. Could you ensure that you use `virtctl` from version `0.22` and ensure that you use `virt-viewer` (just has to be installed on linux or windows)? Maybe that works for you until we have a fix.

**Comment by biliwewe** [comment]
Thank you @rmohr 
Looking forward to fix

**Comment by slintes** [comment]
> Kubevirt upgraded from v0.17.0 to v0.22.0

was the VM restarted? Or is it still running with the old virt-launcher version?

**Comment by biliwewe** [comment]
@slintes yes, I restarted vm. 
View virt-launcher logs
<details>
<summary>kubectl describe pod virt-launcher-vmi-windows1-q87t4:</summary>

```shell
$ kubectl describe pods virt-launcher-win2k3-sqjbt
Name:           virt-launcher-win2k3-sqjbt
Namespace:      default
Priority:       0
Node:           ip-192-168-4-101/192.168.4.101
Start Time:     Wed, 02 Oct 2019 16:52:28 +0800
Labels:         kubevirt.io=virt-launcher
                kubevirt.io/created-by=f69df8e4-e4f1-11e9-b872-0050563d4d0a
Annotations:    cni.projectcalico.org/podIP: 10.42.0.23/32
                kubevirt.io/domain: win2k3
                traffic.sidecar.istio.io/kubevirtInterfaces: k6t-eth0
Status:         Running
IP:             10.42.0.23
IPs:            <none>
Controlled By:  VirtualMachineInstance/win2k3
Containers:
  volumetestc:
    Container ID:  docker://3626b5238b9fe18dbe3aa8b5f75a23b591cb181a5e9c2e6cee9ddcab5e8b7bdf
    Image:         kubevirt/virtio-container-disk
    Image ID:      docker-pullable://kubevirt/virtio-container-disk@sha256:7e5449cb6a4a9586a3cd79433eeaafd980cb516119c03e499492e1e37965fe82
    Port:          <none>
    Host Port:     <none>
    Command:
      /usr/bin/container-disk
    Args:
      --copy-path
      /var/run/kubevirt-ephemeral-disks/container-disk-data/f69df8e4-e4f1-11e9-b872-0050563d4d0a/disk_0
    State:          Running
      Started:      Wed, 02 Oct 2019 16:52:49 +0800
    Ready:          True
    Restart Count:  0
    Limits:
      cpu:     100m
      memory:  40M
    Requests:
      cpu:        10m
      memory:     1M
    Readiness:    exec [/usr/bin/container-disk --health-check] delay=1s timeout=1s period=1s #success=1 #failure=5
    Environment:  <none>
    Mounts:
      /usr/bin from virt-bin-share-dir (rw)
      /var/run/kubevirt-ephemeral-disks/container-disk-data/f69df8e4-e4f1-11e9-b872-0050563d4d0a from container-disks (rw)
  compute:
    Container ID:  docker://c68079a430c8cab3eed718db69b479b9640671e862ad99e83c83c538879645fd
    Image:         index.docker.io/kubevirt/virt-launcher@sha256:1b652f5f99a5a41f450cec878d1cb435020685925f81fb00f3bf9ba7da94e95c
    Image ID:      docker-pullable://kubevirt/virt-launcher@sha256:1b652f5f99a5a41f450cec878d1cb435020685925f81fb00f3bf9ba7da94e95c
    Port:          <none>
    Host Port:     <none>
    Command:
      /usr/bin/virt-launcher
      --qemu-timeout
      5m
      --name
      win2k3
      --uid
      f69df8e4-e4f1-11e9-b872-0050563d4d0a
      --namespace
      default
      --kubevirt-share-dir
      /var/run/kubevirt
      --ephemeral-disk-dir
      /var/run/kubevirt-ephemeral-disks
      --container-disk-dir
      /var/run/kubevirt/container-disks
      --readiness-file
      /var/run/kubevirt-infra/healthy
      --grace-period-seconds
      15
      --hook-sidecars
      0
      --less-pvc-space-toleration
      10
    State:          Running
      Started:      Wed, 02 Oct 2019 16:52:50 +0800
    Ready:          True
    Restart Count:  0
    Limits:
      devices.kubevirt.io/kvm:  1
      devices.kubevirt.io/tun:  1
    Requests:
      cpu:                      100m
      devices.kubevirt.io/kvm:  1
      devices.kubevirt.io/tun:  1
      memory:                   2267208Ki
    Readiness:                  exec [cat /var/run/kubevirt-infra/healthy] delay=4s timeout=5s period=1s #success=1 #failure=5
    Environment:                <none>
    Mounts:
      /var/run/kubevirt from virt-share-dir (rw)
      /var/run/kubevirt-ephemeral-disks from ephemeral-disks (rw)
      /var/run/kubevirt-infra from infra-ready-mount (rw)
      /var/run/kubevirt-private/vmi-disks/pvcdisk from pvcdisk (rw)
      /var/run/kubevirt/container-disks from container-disks (rw)
      /var/run/libvirt from libvirt-runtime (rw)
Conditions:
  Type              Status
  Initialized       True 
  Ready             True 
  ContainersReady   True 
  PodScheduled      True 
Volumes:
  pvcdisk:
    Type:       PersistentVolumeClaim (a reference to a PersistentVolumeClaim in the same namespace)
    ClaimName:  win2k3pvc
    ReadOnly:   false
  infra-ready-mount:
    Type:       EmptyDir (a temporary directory that shares a pod's lifetime)
    Medium:     
    SizeLimit:  <unset>
  virt-share-dir:
    Type:          HostPath (bare host directory volume)
    Path:          /var/run/kubevirt
    HostPathType:  
  virt-bin-share-dir:
    Type:          HostPath (bare host directory volume)
    Path:          /var/lib/kubevirt/init/usr/bin
    HostPathType:  
  libvirt-runtime:
    Type:       EmptyDir (a temporary directory that shares a pod's lifetime)
    Medium:     
    SizeLimit:  <unset>
  ephemeral-disks:
    Type:       EmptyDir (a temporary directory that shares a pod's lifetime)
    Medium:     
    SizeLimit:  <unset>
  container-disks:
    Type:          HostPath (bare host directory volume)
    Path:          /var/run/kubevirt/container-disks/f69df8e4-e4f1-11e9-b872-0050563d4d0a
    HostPathType:  
QoS Class:         Burstable
Node-Selectors:    kubevirt.io/schedulable=true
Tolerations:       node.kubernetes.io/not-ready:NoExecute for 300s
                   node.kubernetes.io/unreachable:NoExecute for 300s
Events:
  Type     Reason     Age                From                       Message
  ----     ------     ----               ----                       -------
  Normal   Scheduled  14m                default-scheduler          Successfully assigned default/virt-launcher-win2k3-sqjbt to ip-192-168-4-101
  Normal   Pulling    14m                kubelet, ip-192-168-4-101  Pulling image "kubevirt/virtio-container-disk"
  Normal   Pulled     14m                kubelet, ip-192-168-4-101  Successfully pulled image "kubevirt/virtio-container-disk"
  Normal   Created    14m                kubelet, ip-192-168-4-101  Created container volumetestc
  Normal   Started    14m                kubelet, ip-192-168-4-101  Started container volumetestc
  Normal   Pulled     14m                kubelet, ip-192-168-4-101  Container image "index.docker.io/kubevirt/virt-launcher@sha256:1b652f5f99a5a41f450cec878d1cb435020685925f81fb00f3bf9ba7da94e95c" already present on machine
  Normal   Created    14m                kubelet, ip-192-168-4-101  Created container compute
  Normal   Started    14m                kubelet, ip-192-168-4-101  Started container compute
  Warning  Unhealthy  14m (x3 over 14m)  kubelet, ip-192-168-4-101  Readiness probe failed: cat: /var/run/kubevirt-infra/healthy: No such file or directory

```
</details>

virt-launcher image info :
```json
[
    {
        "Id": "sha256:b4cd9d00c656e248c343504d888c62908f8fe5e87db4f7c5560e5c36258ec6b3",
        "RepoTags": [],
        "RepoDigests": [
            "kubevirt/virt-launcher@sha256:1b652f5f99a5a41f450cec878d1cb435020685925f81fb00f3bf9ba7da94e95c"
        ],
        "Parent": "",
        "Comment": "",
        "Created": "1970-01-01T00:00:00Z",
        "Container": "",
        "ContainerConfig": {
            "Hostname": "",
            "Domainname": "",
            "User": "",
            "AttachStdin": false,
            "AttachStdout": false,
            "AttachStderr": false,
            "Tty": false,
            "OpenStdin": false,
            "StdinOnce": false,
            "Env": null,
            "Cmd": null,
            "Image": "",
            "Volumes": null,
            "WorkingDir": "",
            "Entrypoint": null,
            "OnBuild": null,
            "Labels": null
        },
        "DockerVersion": "",
        "Author": "Bazel",
        "Config": {
            "Hostname": "",
            "Domainname": "",
            "User": "",
            "AttachStdin": false,
            "AttachStdout": false,
            "AttachStderr": false,
            "Tty": false,
            "OpenStdin": false,
            "StdinOnce": false,
            "Env": [
                "PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin",
                "DISTTAG=f30container",
                "FGC=f30",
                "FBR=f30",
                "container=docker",
                "LIBVIRT_VERSION=5.0.0"
            ],
            "Cmd": [
                "/libvirtd.sh"
            ],
            "ArgsEscaped": true,
            "Image": "sha256:f8b0bdf6cc063c306a4b93de581d7728a1fc3ddb75bb3fa64842a659ffca6ad4",
            "Volumes": null,
            "WorkingDir": "",
            "Entrypoint": [
                "/usr/bin/virt-launcher"
            ],
            "OnBuild": [],
            "Labels": {
                "maintainer": "Clement Verna <cverna@fedoraproject.org>"
            }
        },
        "Architecture": "amd64",
        "Os": "linux",
        "Size": 788709385,
        "VirtualSize": 788709385,
        "GraphDriver": {
            "Data": {
                "LowerDir": "/var/lib/docker/overlay2/b5870694b934415671825fc51e207d88e14f7e8caab30dac29e4455604d63fda/diff:/var/lib/docker/overlay2/e5d1f4760a936433fda511a911e6e830f8b3c3a08217ce4a4126bb838647220a/diff:/var/lib/docker/overlay2/89b903b945273857f44a34e11b6c49c372fa14463f9a885a1772ee9f184b24fe/diff:/var/lib/docker/overlay2/34950377a8ef318479919bc26c185845fadb31183add3de55daab830e8bf6e14/diff:/var/lib/docker/overlay2/c0de1d5806fcc0ade37c590e43ebe47e5542935c24cd61a208fa6b71d5e0370c/diff:/var/lib/docker/overlay2/2691be28c830bc5b4b5d8a1bbd34908c2dcb40471997a3bade565aff9db87be9/diff:/var/lib/docker/overlay2/afb07391ff4712f4f57e114f3011122aaba12adc015d84282d1f00cc5811c528/diff:/var/lib/docker/overlay2/3d8f76ea973eaa62544849d1259399d06ef1b1fc8e6b4f432d3947d86849f310/diff",
                "MergedDir": "/var/lib/docker/overlay2/f67ea588c115a20496e6a71aba8311f9c22d09a67eb3dd25a6bae763725eb02c/merged",
                "UpperDir": "/var/lib/docker/overlay2/f67ea588c115a20496e6a71aba8311f9c22d09a67eb3dd25a6bae763725eb02c/diff",
                "WorkDir": "/var/lib/docker/overlay2/f67ea588c115a20496e6a71aba8311f9c22d09a67eb3dd25a6bae763725eb02c/work"
            },
            "Name": "overlay2"
        },
        "RootFS": {
            "Type": "layers",
            "Layers": [
                "sha256:253342a51b7a3d0f01018214663173ed8dc9839f84bb11d28c664611749dd799",
                "sha256:389d292df5bf2326e7bd17eee07d66d0a4b21b0c988cebbd965339316ecf1a0d",
                "sha256:3329aecb1dfb8ab8c2c7ef92d96a185bdc398b4755be6975bea02d3c6081e110",
                "sha256:7595d9c9db6fce04ccb747908cda1781b08823b1fc99f2f9a659b4d9b66e1c90",
                "sha256:09e0eb6267a087e0c8543614122a2c3c5c174aa61adffdc4a57f07e71fe0b705",
                "sha256:fdd653aa63fa2ba2ee79902123d8db92a527415e82953d80e46e417753f4854e",
                "sha256:5b7a2c9db28b852249fa513162d883a816d3877b129ccf7abe09922fb3925fb7",
                "sha256:4384d578d510ef7e376efe20d1faa42e2b791d89a7da920b5ec4108a0da5f06b",
                "sha256:c43f6b0d6abdd3f076801c3cb8ff7e352583176a2753f029c978266a749eb481"
            ]
        },
        "Metadata": {
            "LastTagTime": "0001-01-01T00:00:00Z"
        }
    }
]

```
I also tried v0.19.0 and v0.21.0 versions. Only v.0.17.0 version will not show bugs
https://github.com/kubevirt/kubevirt/issues/2539#issue-473828756

**Comment by slintes** [comment]
@rmohr was investigating this further, and figured out that it might have something to do with proxy servers (see #virtualization channel in the k8s slack). Do you use a proxy?

**Comment by rmohr** [comment]
Finally have a fix in #2872. Need to clean it up a little, but should be in soon.

**Comment by biliwewe** [comment]
@rmohr Thank you
